x-pack/legacy/plugins/monitoring/index.ts

```diff
@@ -0,0 +1,150 @@
+
+import { resolve } from 'path';
+import { Server } from 'src/legacy/server/kbn_server';
+import KbnServer from 'src/legacy/server/kbn_server';
+import { i18n } from '@kbn/i18n';
+
+import { LegacyPluginApi, LegacyPluginSpec, ArrayOrItem } from 'src/legacy/plugin_discovery/types';
+
+// @ts-ignore
+import { config as configDefaults } from './config';
+// @ts-ignore
+import { deprecations } from './deprecations';
+// @ts-ignore
+import { Plugin } from './server/plugin';
+// @ts-ignore
+import { initInfraSource } from './server/lib/logs/init_infra_source';
+
+type InfraPlugin = any; // TODO
+type PluginsSetup = any; // TODO
+type LegacySetup = any; // TODO
+
+const uiExports = {
+  app: {
+    title: i18n.translate('xpack.monitoring.stackMonitoringTitle', {
+      defaultMessage: 'Stack Monitoring',
+    }),
+    order: 9002,
+    icon: 'plugins/monitoring/icons/monitoring.svg',
+    euiIconType: 'monitoringApp',
+    main: 'plugins/monitoring/legacy',
+  },
+  injectDefaultVars(server: Server) {
+    const config = server.config();
+    return {
+      // TODO: Implement with PluginConfigDescriptor.exposeToBrowser instead
+      showLicenseExpiration: config.get('xpack.monitoring.show_license_expiration'),
+      monitoringUiEnabled: config.get('xpack.monitoring.ui.enabled'),
+      minIntervalSeconds: config.get('xpack.monitoring.min_interval_seconds'),
+      showCgroupMetricsElasticsearch: config.get(
+        'xpack.monitoring.ui.container.elasticsearch.enabled'
+      ),
+      showCgroupMetricsLogstash: config.get('xpack.monitoring.ui.container.logstash.enabled'),
+    };
+  },
+  hacks: ['plugins/monitoring/hacks/toggle_app_link_in_nav'],
+  home: ['plugins/monitoring/register_feature'],
+  styleSheetPaths: resolve(__dirname, 'public/index.scss'),
+};
+
+const validConfigOptions: string[] = [
+  'xpack.monitoring.ui.enabled',
+  'xpack.monitoring.kibana.collection.enabled',
+  'xpack.monitoring.max_bucket_size',
+  'xpack.monitoring.min_interval_seconds',
+  'kibana.index',
+  'xpack.monitoring.show_license_expiration',
+  'xpack.monitoring.ui.container.elasticsearch.enabled',
+  'xpack.monitoring.ui.container.logstash.enabled',
+  'xpack.monitoring.tests.cloud_detector.enabled',
+  'xpack.monitoring.kibana.collection.interval',
+  'xpack.monitoring.elasticsearch.hosts',
+  'xpack.monitoring.elasticsearch',
+  'xpack.monitoring.xpack_api_polling_frequency_millis',
+  'server.uuid',
+  'server.name',
+  'server.host',
+  'server.port',
+  'xpack.monitoring.cluster_alerts.email_notifications.enabled',
+  'xpack.monitoring.cluster_alerts.email_notifications.email_address',
+  'xpack.monitoring.ccs.enabled',
+  'xpack.monitoring.elasticsearch.logFetchCount',
+];
+
+/**
+ * Invokes plugin modules to instantiate the Monitoring plugin for Kibana
+ * @param kibana {Object} Kibana plugin instance
+ * @return {Object} Monitoring UI Kibana plugin object
+ */
+export const monitoring = (kibana: LegacyPluginApi): ArrayOrItem<LegacyPluginSpec> => {
```

Comment:
The usage of `ArrayOrItem` type confuses me? Would this function ever return an array?

Reply:
You're right, looks like the type allows for both `T` and `T[]` I will change it to a partial `LegacyPluginSpec`, since we're only returning one plugin